Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/SharedKeyHttpObservable.java

```diff
@@ -21,40 +21,106 @@
+import rx.Observable;
+import rx.exceptions.Exceptions;
 
-import java.security.InvalidKeyException;
+import java.io.IOException;
+import java.net.UnknownServiceException;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.List;
+import java.util.Optional;
 import java.util.UUID;
 
-public class SharedKeyHttpObservable extends HttpObservable {
+public class SharedKeyHttpObservable extends AzureHttpObservable {
     public static String ApiVersion = "2018-11-09";
     public static String DefaultContentLength = "0";
+    public static String AccessKeyUri = "https://management.azure.com/api/invoke";
     private SharedKeyCredential cred;
 
     public SharedKeyHttpObservable(String accountName, String accessKey) {
+        super(ApiVersion);
+        initHeaderGroupForFSOperations();
+        try {
+            this.cred = new SharedKeyCredential(accountName, accessKey);
+        } catch (IllegalArgumentException ex) {
+            log().warn("Create shared key credential encounter exception", ex);
+            throw new IllegalArgumentException("Can't create shared key credential.Please check access key");
+        }
+    }
+
+    public SharedKeyHttpObservable(SubscriptionDetail subscription, String accountName) {
+        super(subscription, ApiVersion);
+        getAccessKey(subscription, accountName)
+                .subscribe(resp -> {
+                            try {
+                                Optional<AccessKeyList> keys = ObjectConvertUtils.convertEntityToObject(resp.getEntity(), AccessKeyList.class);
+                                if (keys.isPresent()) {
+                                    this.cred = new SharedKeyCredential(accountName, keys.get().getKeys().get(0).getValue());
+
+                                } else {
+                                    throw new UnknownServiceException("Can't get valid access key from response");
+                                }
+                            } catch (IOException ex) {
+                                log().warn("Create shared key credential encounter exception", ex);
+                                throw new IllegalArgumentException("Can't create shared key credential.Please check key validation");
+                            }
+                        }
+                );
+
+        initHeaderGroupForFSOperations();
+    }
+
+    private void initHeaderGroupForFSOperations() {
         HeaderGroup defaultHeaders = new HeaderGroup();
         defaultHeaders.addHeader(new BasicHeader("x-ms-client-request-id", UUID.randomUUID().toString()));
         defaultHeaders.addHeader(new BasicHeader("x-ms-date", Utility.getGMTTime()));
         defaultHeaders.addHeader(new BasicHeader("x-ms-version", ApiVersion));
         defaultHeaders.addHeader(new BasicHeader("authorization", ""));
         defaultHeaders.addHeader(new BasicHeader("Content-Type", "application/json"));
-
         setDefaultHeaderGroup(defaultHeaders);
```

Comment:
>setDefaultHeaderGroup [](start = 8, length = 21)

Why override the default header group?

Reply:
generating shared key dosen't need the original headers and can have double authorization headers

---
In reply to: [271138320](https://github.com/Microsoft/azure-tools-for-java/pull/2961#discussion_r271138320) [](ancestors = 271138320)